### PR TITLE
[ONNX] Fix numerical errors in softmax when dim is not last dimension

### LINKED ISSUE #34585: Unstable softmax trace when exporting to onnx
## 🐛 Bug

The ONNX model with softmax which is exported with torch.onnx module predicts NaNs instead of expected numerical values. The actual output does not depend on mantissa. The actual result is unstable. Meaning that expected value 5.0390004e-32 actually may be converted either to NaN or to 5.0390004e-32.  If for example 5.0390004e-32 was correctly converted to  5.0390004e-32 then it may happen that other element of the tensor, e.g. 1.0000000e+00, might be converted to NaN.

## To Reproduce
Please consider the following code to reproduse
```
import torch.onnx
import torch.nn as nn
import onnxruntime as rt
import torch.nn.functional as F


class TraceCheck(nn.Module):
    def __init__(self):
        super(TraceCheck, self).__init__()

    def forward(self, x):
        return F.softmax(-100*x, -2)


if __name__ == '__main__':
    net = TraceCheck()
    x = torch.Tensor(torch.randn(5,5))
    pred = net.forward(x).detach().cpu().numpy()
    tmp_filename = 'test.onnx'
    torch.onnx.export(net, x, tmp_filename)
    sess = rt.InferenceSession(tmp_filename)
    pred_onnx = sess.run([sess.get_outputs()[-1].name], {sess.get_inputs()[-1].name: x.detach().cpu().numpy()})[0]
    print(pred)
    print(pred_onnx)
```

Example outputs are :
pred:
>[[1.0000000e+00 0.0000000e+00 0.0000000e+00 1.0000000e+00 0.0000000e+00]
 >[0.0000000e+00 0.0000000e+00 0.0000000e+00 0.0000000e+00 0.0000000e+00]
 >[0.0000000e+00 6.1350977e-01 0.0000000e+00 0.0000000e+00 0.0000000e+00]
 >[0.0000000e+00 3.8649023e-01 8.0113404e-19 0.0000000e+00 1.0000000e+00]
 >[1.1149160e-09 5.0298510e-32 1.0000000e+00 0.0000000e+00 0.0000000e+00]]

pred_onnx:
>[[          nan 0.0000000e+00 0.0000000e+00           nan 0.0000000e+00]
> [0.0000000e+00 0.0000000e+00 0.0000000e+00 0.0000000e+00 0.0000000e+00]
>[0.0000000e+00 6.1350977e-01 0.0000000e+00 0.0000000e+00 0.0000000e+00]
>[0.0000000e+00 3.8649023e-01           nan 0.0000000e+00           nan]
>[          nan 5.0298498e-32           nan 0.0000000e+00 0.0000000e+00]]

We see that pred != pred_onnx

## Expected behavior

pred == pred_onnx

## Environment

- PyTorch Version : 1.4.0
- OS: Windows 10 Pro 64 bit
- How you installed PyTorch (conda, pip, source): current repo (git clone ...)
- Build command you used (if compiling from source): setup.py build/install
- Python version: 3.6
- CUDA/cuDNN version: None
- GPU models and configuration: None
- Any other relevant information: opset 11

## Additional context

cc @houseroad @spandantiwari @lara-hdr @BowenBao @neginraoof

### PR DESCRIPTION
**Pull request by p12tic**
Fixes #34585.

This PR improves the workaround for the problem of different semantics between ONNX softmax and Pytorch softmax.

In Pytorch the `dim` parameter specifies over which dimension normalize the values.  ONNX on the other hand always coerces the input into a 2D tensor and the `axis` parameter specifies which dimensions represent rows and columns of the resulting tensor. As a result, only when we are normalizing the last dimension (`dim == ndim - 1`) semantics are the same.

Previously this was handled by recognizing the `dim == ndim - 1` case and using `softmax` for that. All other cases used a fallback path of explicit invocations of exp, reducesum and div operators to compute the result. Unfortunately, this results in numeric errors when input values are large: the result of exp will produce infinity on both numerator and denumerator and the division of that will result in NaN.

This can be improved by transposing the input tensor so that we can reuse ONNX softmax.

Similar approach has been applied to `logsoftmax` function in #30433.

### CODE REVIEW

File: test/onnx/test_pytorch_onnx_onnxruntime.py [hunk 1]
```diff
@@ -1648,6 +1648,12 @@ def forward(self, input):
         x = torch.randn(4, 4, requires_grad=True)
         self.run_test(ReduceLogSumExpModel(), x)
 
+    def test_softmax(self):
+        for i in range(-4, 3):
+            model = torch.nn.Softmax(dim=i)
+            input = torch.randn(3, 4, 5, 6)
+            self.run_test(model, input)
+
```
Comment (BowenBao) [inline test/onnx/test_pytorch_onnx_onnxruntime.py:1656]: Please add the following test case for large values
```python
    def test_softmax_large_values(self):
        input = torch.tensor([[-1e12, -1e12, -1e12], [1e12, 0.0, -5.0], [3.0, 4.0, 5.0]])
        for i in range(-2, 1):
            model = torch.nn.Softmax(dim=0)
            self.run_test(model, input)

            class SoftmaxUnknownRank(torch.nn.Module):
                def __init__(self, i):
                    super().__init__()
                    self.softmax = torch.nn.Softmax(dim=i)

                def forward(self, x):
                    return self.softmax(x.reshape(3, 3))

            model = torch.jit.script(SoftmaxUnknownRank(i))
            self.run_test(model, input)
```

File: torch/onnx/symbolic_opset9.py [hunk 2]
```diff
@@ -584,19 +584,32 @@ def softmax(g, input, dim, dtype=None):
...
+            dim = input_dim - 1
+
+        softmax = g.op('Softmax', input, axis_i=dim)
+        if dtype and dtype.node().kind() != 'prim::Constant':
+            parsed_dtype = sym_help._get_const(dtype, 'i', 'dtype')
+            softmax = g.op("Cast", softmax, to_i=sym_help.scalar_type_to_onnx[parsed_dtype])
+
+        if is_transpose_required:
+            softmax = g.op("Transpose", softmax, perm_i=axes)
+        return softmax
```
Comment (BowenBao) [inline torch/onnx/symbolic_opset9.py:613]: please add max normalization here for support of cases where input rank is unknown.
```python
    # Apply max normalization.
    input = g.op('Sub', input, g.op('ReduceMax', input, axes_i=[dim], keepdims_i=1))
```

**Review by BowenBao — CHANGES_REQUESTED** [review]
Thanks @p12tic for the fix! Overall looks good, please see inline comments for some additional updates.

**Comment by BowenBao** [comment]
@p12tic could you take a look at the comments and update your PR? Thanks!

**Comment by p12tic** [comment]
@BowenBao Yes, I will do so. This week my priorities were elsewhere, thus the delay.

File: test/onnx/test_pytorch_onnx_onnxruntime.py (hunk 1 above)
Comment (p12tic) [reply test/onnx/test_pytorch_onnx_onnxruntime.py:1679]: Done, thank you! I also added SoftmaxUnknownRank case for the non-large values test.

File: torch/onnx/symbolic_opset9.py (hunk 2 above)
Comment (p12tic) [reply torch/onnx/symbolic_opset9.py:613]: Done.

**Comment by p12tic** [comment]
@BowenBao I've applied your suggestions to the PR. Thanks for suggesting them to the PR instead of doing it yourself, I've got to learn what max normalization in softmax is :-)

**Review by BowenBao — APPROVED** [review]
@p12tic Thanks for the fix! LGTM
